### ISSUE TITLE
[models] Training pipeline refactoring

### ISSUE THREAD

**Issue by dataJSA** [pyronear/pyro-risks#36]
Currently, the training module `score_v0` contains the two model training functions and helper functions that were developed for fitting models before the demo day. In order to improve the modularity and the extensibility of the code base it would be great to refactor the `pyro_risks.models` modules. 

# Design

For refactoring I propose using imbalanced-learn / scikit-learn Pipeline classes and  implementing the different steps as transformers.

The pipelines are a pretty neat way to define a sequence of data preprocessing and modeling steps. Especially the imbalanced-learn pipelines they allow for **defining sequences of resampling, preprocessing and modeling steps as one estimators**. 

- Here is some documentation on the different pipelines [datasets transformation](https://scikit-learn.org/stable/data_transforms.html)
- Here is the suggestion for the `pyro-risks.models` sub-package and the root modules

    ```bash
    ├── config
    │   ├── __init__.py
    │   ├── datasets.py
    │   ├── logger.py
    │   └── models.py
    ├── models
    │   ├── __init__.py
    │   ├── pipelines.py
    │   ├── transformers.py
    │   └── utils.py
    ├── predict.py
    ├── train.py
    └── version.py
    ```

# Benefits

I believe the use of training pipelines will have many benefits, to name a few using pipelines will help:

- Ensure consistency between pipeline implementations
- Ensure consistency between training and prediction
- Improve hyper-parameter tuning traceability
- Improve training reproducibility
- Improve pipeline readability
- Avoiding data leakage

⭐ Bonus ⭐ 
I believe that further relying on sklearn ecosystem, the tools that they provide and  their documentation will help improve new contributors onboarding and help us build on each other works.

# Additional context 

Since the current training and prediction scripts share common functions, the refactoring scope (training + prediction) is quite large. I propose dividing it into two separate issues and pull request, leaving the previous training functions in the mean time for ensuring backward compatibility.

Let me know what you think guys !

**Comment by GHCamille** [comment]
I agree on the benefits pipelines, and on making two issues since it's not totally the same work. 

I just have one question about imbalanced-learn, where do you think it would be in the pipeline ? I have trouble understanding it it would be at the beginning/after preprocessing/to test new sampling techniques ?

**Comment by dataJSA** [comment]
Ok I see what you mean, my first guest  would be to define the pipeline steps in the following order:
- preprocessing >> upsampling >> modeling  

otherwise we could skew the training data resulting for instance in a less accurate imputation of missing values.

Also, the imbalanced-learn pipeline implementation is really great since it abstract away some of the difficulties of using resampling (avoiding resampling the test set and validation sets,  we can be complicated for complexe cross validation techniques). 

The following article demonstrate how to set up a pipeline imbalanced-learn [SMOTE for Imbalanced Classification with Python](https://machinelearningmastery.com/smote-oversampling-for-imbalanced-classification/)
